### PR TITLE
[Validator] add new `Timezone` validation constraint

### PR DESCRIPTION
**Pull request by hhamon**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/11317

Rework of #22262.

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ */
+class TimezoneValidator extends ConstraintValidator
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof Timezone) {
+            throw new UnexpectedTypeException($constraint, __NAMESPACE__.'\Timezone');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: `Timezone::class`

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    protected function formatValue($value, $format = 0)
+    {
+        $value = parent::formatValue($value, $format);
+
+        if (!$value || \DateTimeZone::PER_COUNTRY === $value) {
+            return $value;
+        }
+
+        if ($zoneFound = array_search($value, (new \ReflectionClass(\DateTimeZone::class))->getConstants(), true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: `return array_search($value, (new \ReflectionClass(\DateTimeZone::class))->getConstants(), true) ?: $value;`

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php (hunk 1 above)
Comment (hhamon) [reply src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: Fixed.

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php (hunk 2 above)
Comment (hhamon) [reply src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: Fixed.

**Comment by hhamon** [comment]
Also documentation PR is on its way as well ;)

File: src/Symfony/Component/Validator/Resources/translations/validators.fr.xlf [hunk 3]
```diff
@@ -334,6 +334,10 @@
                 <source>This value should be valid JSON.</source>
                 <target>Cette valeur doit être un JSON valide.</target>
             </trans-unit>
+            <trans-unit id="92">
+                <source>This value is not a valid timezone.</source>
+                <target>Cette valeur n'est pas un identifiant de fuseau horaire valide.</target>
```
Comment (jewome62) [inline src/Symfony/Component/Validator/Resources/translations/validators.fr.xlf]: `Cette valeur n'est pas un fuseau horaire valide` ?
Into other translation, you do not speak about identifier. Just in French

Comment (hhamon) [reply src/Symfony/Component/Validator/Resources/translations/validators.fr.xlf]: Fixed. thanks.

File: src/Symfony/Component/Validator/Resources/translations/validators.en.xlf [hunk 4]
```diff
@@ -354,6 +354,10 @@
                 <source>This collection should contain only unique elements.</source>
                 <target>This collection should contain only unique elements.</target>
             </trans-unit>
+            <trans-unit id="92">
+                <source>This value is not a valid timezone.</source>
+                <target>This value is not a valid timezone.</target>
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Resources/translations/validators.en.xlf]: Dieser Wert ist keine gültige Zeitzone.

For the German file

Comment (hhamon) [reply src/Symfony/Component/Validator/Resources/translations/validators.en.xlf]: Great thanks :)

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        if (!$constraint instanceof Timezone) {
+            throw new UnexpectedTypeException($constraint, Timezone::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedTypeException($value, 'string');
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: should be `UnexpectedValueException`

**Review by xabbuh — APPROVED** [review]
with a minor change request

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php (hunk 5 above)
Comment (hhamon) [reply src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: Fixed!
Thanks!

File: src/Symfony/Component/Validator/Constraints/Timezone.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * @Annotation
+ * @Target({"PROPERTY", "METHOD", "ANNOTATION"})
+ *
+ * @author Javier Spagnoletti <phansys@gmail.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ */
+class Timezone extends Constraint
+{
+    public const NO_SUCH_TIMEZONE_IDENTIFIER_ERROR = '5ce113e6-5e64-4ea2-90fe-d2233956db13';
```
Comment (fabpot) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: Can we make the `const`s shorter? Like removing the `NO_SUCH_` prefix?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Constraints/Timezone.php (hunk 6 above)
Comment (hhamon) [reply src/Symfony/Component/Validator/Constraints/Timezone.php]: Prefixes have been removed.

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        } else {
+            $timezoneIds = \DateTimeZone::listIdentifiers($constraint->zone);
+        }
+
+        if ($timezoneIds && \in_array($value, $timezoneIds, true)) {
+            return;
+        }
+
+        if ($constraint->countryCode) {
+            $code = Timezone::NO_SUCH_TIMEZONE_IDENTIFIER_IN_COUNTRY_ERROR;
```
Comment (fabpot) [inline src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: Looks like you need to remove the prefix everywhere :)

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Constraints/TimezoneValidator.php (hunk 7 above)
Comment (hhamon) [reply src/Symfony/Component/Validator/Constraints/TimezoneValidator.php]: Oops! Thanks :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @hhamon.

**Comment by ro0NL** [comment]
do we still consider https://github.com/symfony/symfony/pull/22262#pullrequestreview-179791874?

**Pull request merged**